src/dbnode/storage/index.go

```diff
@@ -650,18 +651,22 @@ func (i *nsIndex) writeBatches(
 				return
 			}
 
-			if !ts.After(pastLimit) {
+			if ts.Before(pastLimit) {
 				// NB(bodu): We only mark entries as too far in the past if
 				// cold writes are not enabled.
-				var err error
 				if !i.coldWritesEnabled {
-					err = m3dberrors.ErrTooPast
+					batch.MarkUnmarkedEntryError(m3dberrors.ErrTooPast, idx)
+					return
+				}
+				// We allow index writes that are between mutable block start and
+				// start of buffer past to go through.
+				if ts.Before(warmBlockStart) {
+					// NB(bodu): When we mark a index cold write with a <nil> error
+					// we are recording an indexing attempt for this series and will not
+					// attempt to index the series again. This is correct for cold index writes
+					// because they are not handled in warm index writes path.
+					batch.MarkUnmarkedEntryError(nil, idx)
```

Comment:
Hm, think you want to move the `return` into this sub-`if` branch otherwise forward indexing and the `notSkipped` var does not get incremented.